File: 12_Heaps/M-InterviewBit-Maximum Sum Combination.java
// Maximum Sum Combination
// https://www.interviewbit.com/problems/maximum-sum-combinations/

// Problem Statement:
// Given two arrays A and B of size N each. Find the maximum N elements from the sum combinations (Ai + Bj) formed from elements in array A and B.
// Example:
// Input: A = [1, 4, 2, 3], B = [2, 5, 1, 6], N = 4
// Output: [10, 9, 9, 8]
// Explanation: The maximum 4 elements of the sum combinations are 10(6+4), 9(6+3), 9(5+4), 8(6+2).

import java.util.*;

class Solution {

    public int[] solve(int[] A, int[] B, int C) {
        int n = A.length;

        // Sort both arrays in descending order
        Arrays.sort(A);
        Arrays.sort(B);

        // Reverse both arrays to get descending order
        reverse(A);
        reverse(B);

        // Max heap to store sums with their indices
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> b[0] - a[0]);
        Set<String> visited = new HashSet<>();

        // Start with the maximum possible sum
        pq.offer(new int[] { A[0] + B[0], 0, 0 });
        visited.add("0,0");

        int[] result = new int[C];
        int index = 0;

        while (index < C && !pq.isEmpty()) {
            int[] curr = pq.poll();
            result[index++] = curr[0];
            int i = curr[1];
            int j = curr[2];

            // Add next possible combinations
            if (i + 1 < n && !visited.contains((i + 1) + "," + j)) {
                pq.offer(new int[] { A[i + 1] + B[j], i + 1, j });
                visited.add((i + 1) + "," + j);
            }

            if (j + 1 < n && !visited.contains(i + "," + (j + 1))) {
                pq.offer(new int[] { A[i] + B[j + 1], i, j + 1 });
                visited.add(i + "," + (j + 1));
            }
        }

        return result;
    }

    private void reverse(int[] arr) {
        int left = 0,
            right = arr.length - 1;
        while (left < right) {
            int temp = arr[left];
            arr[left] = arr[right];
            arr[right] = temp;
            left++;
            right--;
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // Test case 1: Basic example
        int[] A1 = { 1, 4, 2, 3 };
        int[] B1 = { 2, 5, 1, 6 };
        int C1 = 4;
        int[] result1 = solution.solve(A1, B1, C1);
        System.out.println("Test 1 - Expected: [10, 9, 9, 8]");
        System.out.print("Test 1 - Actual: [");
        for (int i = 0; i < result1.length; i++) {
            System.out.print(result1[i]);
            if (i < result1.length - 1) System.out.print(", ");
        }
        System.out.println("]");

        // Test case 2: Small identical arrays
        int[] A2 = { 5, 3, 1 };
        int[] B2 = { 5, 3, 1 };
        int C2 = 3;
        int[] result2 = solution.solve(A2, B2, C2);
        System.out.println("Test 2 - Expected: [10, 8, 8]");
        System.out.print("Test 2 - Actual: [");
        for (int i = 0; i < result2.length; i++) {
            System.out.print(result2[i]);
            if (i < result2.length - 1) System.out.print(", ");
        }
        System.out.println("]");

        // Test case 3: Edge case with maximum values
        int[] A3 = { 1000, 999, 998 };
        int[] B3 = { 1000, 999, 998 };
        int C3 = 5;
        int[] result3 = solution.solve(A3, B3, C3);
        System.out.println("Test 3 - Expected: [2000, 1999, 1999, 1998, 1998]");
        System.out.print("Test 3 - Actual: [");
        for (int i = 0; i < result3.length; i++) {
            System.out.print(result3[i]);
            if (i < result3.length - 1) System.out.print(", ");
        }
        System.out.println("]");
    }
}

File: JavaScript/12_Heaps/M-InterviewBit-Maximum Sum Combination.js
// Maximum Sum Combination
// https://www.interviewbit.com/problems/maximum-sum-combinations/

// Problem Statement:
// Given two arrays A and B of size N each. Find the maximum N elements from the sum combinations (Ai + Bj).

/**
 * MaxHeap implementation for sum combinations
 */
function createMaxHeap() {
    const heap = [];

    function size() {
        return heap.length;
    }

    function isEmpty() {
        return heap.length === 0;
    }

    function push(item) {
        heap.push(item);
        bubbleUp(heap.length - 1);
    }

    function pop() {
        if (heap.length === 0) return undefined;
        const max = heap[0];
        const last = heap.pop();
        if (heap.length > 0) {
            heap[0] = last;
            bubbleDown(0);
        }
        return max;
    }

    function bubbleUp(idx) {
        while (idx > 0) {
            const parent = Math.floor((idx - 1) / 2);
            if (heap[parent][0] >= heap[idx][0]) break;
            [heap[parent], heap[idx]] = [heap[idx], heap[parent]];
            idx = parent;
        }
    }

    function bubbleDown(idx) {
        const n = heap.length;
        while (true) {
            let largest = idx;
            const left = 2 * idx + 1;
            const right = 2 * idx + 2;
            if (left < n && heap[left][0] > heap[largest][0]) largest = left;
            if (right < n && heap[right][0] > heap[largest][0]) largest = right;
            if (largest === idx) break;
            [heap[largest], heap[idx]] = [heap[idx], heap[largest]];
            idx = largest;
        }
    }

    return { size, isEmpty, push, pop };
}

/**
 * Find maximum C sum combinations from arrays A and B
 * @param {number[]} A - First array
 * @param {number[]} B - Second array
 * @param {number} C - Number of maximum sums to find
 * @returns {number[]} - Array of C maximum sums
 */
function solve(A, B, C) {
    const n = A.length;

    // Sort both arrays in descending order
    A.sort((a, b) => b - a);
    B.sort((a, b) => b - a);

    // Max heap to store [sum, i, j]
    const pq = createMaxHeap();
    const visited = new Set();

    // Start with the maximum possible sum
    pq.push([A[0] + B[0], 0, 0]);
    visited.add("0,0");

    const result = [];

    while (result.length < C && !pq.isEmpty()) {
        const [sum, i, j] = pq.pop();
        result.push(sum);

        // Add next possible combinations
        if (i + 1 < n && !visited.has(`${i + 1},${j}`)) {
            pq.push([A[i + 1] + B[j], i + 1, j]);
            visited.add(`${i + 1},${j}`);
        }

        if (j + 1 < n && !visited.has(`${i},${j + 1}`)) {
            pq.push([A[i] + B[j + 1], i, j + 1]);
            visited.add(`${i},${j + 1}`);
        }
    }

    return result;
}

// Test cases
console.log(solve([1, 4, 2, 3], [2, 5, 1, 6], 4)); // [10, 9, 9, 8]
console.log(solve([5, 3, 1], [5, 3, 1], 3)); // [10, 8, 8]
console.log(solve([1, 2], [3, 4], 2)); // [6, 5]
